tests-arquillian/src/test/java/org/jboss/weld/tests/internal/contructs/BeansImplementWeldInternalConstructsTest.java

```diff
@@ -0,0 +1,145 @@
+/*
+ * JBoss, Home of Professional Open Source
+ * Copyright 2017, Red Hat, Inc., and individual contributors
+ * by the @authors tag. See the copyright.txt in the distribution for a
+ * full listing of individual contributors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jboss.weld.tests.internal.contructs;
+
+import javax.enterprise.inject.spi.Bean;
+import javax.enterprise.inject.spi.BeanManager;
+import javax.inject.Inject;
+
+import org.jboss.arquillian.container.test.api.Deployment;
+import org.jboss.arquillian.junit.Arquillian;
+import org.jboss.shrinkwrap.api.Archive;
+import org.jboss.shrinkwrap.api.BeanArchive;
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.weld.proxy.WeldClientProxy;
+import org.jboss.weld.proxy.WeldConstruct;
+import org.jboss.weld.test.util.Utils;
+import org.junit.Assert;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+
+/**
+ *
+ * @author <a href="mailto:manovotn@redhat.com">Matej Novotny</a>
+ */
+@RunWith(Arquillian.class)
+public class BeansImplementWeldInternalConstructsTest {
+
+    @Deployment
+    public static Archive getDeployment() {
+        return ShrinkWrap.create(BeanArchive.class, Utils.getDeploymentNameAsHash(BeansImplementWeldInternalConstructsTest.class))
+            .addPackage(BeansImplementWeldInternalConstructsTest.class.getPackage());
+    }
+
+    @Inject
+    BeanHolder holder;
+
+    @Inject
+    BeanManager bm;
+
+    @Test
+    public void testClientProxyBean() {
+        ClientProxyBean clientProxyBean = holder.getClientProxyBean();
+        // trigger proxy creation
+        clientProxyBean.ping();
+
+        // injected bean should be instance of WeldConstruct and WeldClientProxy
+        Assert.assertTrue(clientProxyBean instanceof WeldConstruct);
+        Assert.assertTrue(clientProxyBean instanceof WeldClientProxy);
+
+        // cast to WeldClientProxy and test the methods
+        WeldClientProxy wcp = (WeldClientProxy) clientProxyBean;
+        
+        Object contextualInstance = wcp.getContextualInstance();
+        // kind of indirect check that this is the actual contextual instance
+        Assert.assertTrue(contextualInstance instanceof ClientProxyBean);
+        Assert.assertFalse(contextualInstance instanceof WeldConstruct);
+        
+        // TODO how to verify this
+        Bean<?> bean = wcp.getBean();
```

Comment:
Just get the bean for `ClientProxyBean` from `BeanManager`, cast to `WeldBean` and compare the bean identifiers?